### DISCUSSION TITLE
Triggers aren't business data constraints

### DISCUSSION THREAD

**Discussion by yrashk** [omnigres/omnigres#705] [Ideas]
We often use triggers to enforce global data constraints (beyond the scope of a single relation) in business systems. However, it is a mechanical representation of "what to do" of something that is higher level, and it has a few problems, in no specific order:

1. Establishing triggers involves ceremony: we need a trigger and a function, so we need to establish their conditions, properties, and a (consistent) naming scheme.
2. We need to write similar-but-not-the-same triggers when changes are brought from any of the relations involved in a business constraint: say, if we have users and currencies relation, and every user must have an account in each, we need to establish triggers on changes in users _and_ currencies. 
3. We can't tell what the rules are by observing triggers, as those are implemented in a non-declarative way.
4. We can't easily tell whether the rules are implemented correctly as, by the nature of the implementation, they need to be tested in different scenarios involving changes in different relations. The more relations are involved, the harder it is.

So here's what I was thinking about recently: can we establish a concept of declarative database-wide constraints to implement such business rules? On the surface, this may seem like the _almost-nobody-implemented-these_ SQL assertions, but this must be thought through more deeply. 

What if we were able to establish a query that would be executed when either of the involved relations got mutably accessed and depending on the logic, it would result in one of these:

1. Passed constraint, no effect.
2. Failed constraint, error.
3. Failed constraint and compensation action.

The case in **(2)** is important to establish some ground rules so that when the database schema is operated, we always know it is in the correct state.
The case in **(3)** is most interesting because it allows us to avoid writing those triggers I described above. 

I am still a bit unclear on how (3) should look like and how it should be implemented, but that's what this discussion is for 😂. At least conceptually, what if the query was a CTE where inserts/updates/delete  would act on sub-queries that are checking for failed constraint and return data sufficient to compensate?

**Comment by danolivo** [comment]
I am not sure why you think this approach could allow you to avoid all the complexity? That is a reason assertions are not yet implemented in Postgres.

**Reply by ebellani** [reply]
I think assert would allow one to use a declarative language to establish an invariant (say, of a proportion of records between 2 tables). Without something like that, one would need to establish several triggers on both tables, using a procedural language at that.

**Comment by danolivo** [comment]
After second thoughts, I do recommend having a cup of tea with C.J. Date. He has a long-term opinion on that subject, and you can discuss it in a more factual manner. I would like to read the transcript of that talk ;)

**Reply by yrashk** [reply]
Not meaning this to be a replacement for that conversation (or be hallucination-free), but here's the AI overlord's summary on what may be C.J. Date's opinion – that can be useful to entertain us (I am in a moving vehicle right now anyway):

<blockquote>
C.J. Date, a prominent relational database theorist, has discussed the importance of comprehensive constraint mechanisms within a relational database system in numerous writings. While he may not have provided extensive commentary on the exact concept of “database-wide constraints” as implemented in modern systems, his views on integrity constraints and relational theory closely align with the idea.

**Context from Date’s Work:**

1. **Assertions and Integrity Constraints:**  
   In the classic relational model—and as presented in much of Date’s work—database integrity is not meant to be confined to single tables. Date has consistently emphasized that relational integrity should be defined at the logical level of the entire database. He has lamented that mainstream SQL implementations do not support the `ASSERTION` feature proposed in the SQL standard. SQL ASSERTIONS were intended to define global integrity constraints—conditions that cross relational boundaries. Date has often noted that the absence of such a feature in commercial products is a significant shortfall from a theoretical standpoint.

2. **Domain and Business Rules:**  
   Date, along with Hugh Darwen and others, argues that business rules (invariants) belong logically in the database. According to relational theory, a database should enforce all aspects of logical integrity consistently and centrally. This includes:
   - **Domain Constraints:** Ensuring each attribute’s values conform to defined sets or domains.
   - **Relational Constraints (Keys, Foreign Keys):** Ensuring correct relationships between tuples in one or multiple relations.
   - **More Complex Constraints:** Enforcing semantic rules that cannot be expressed merely as local checks on single relations. Date sees the database as the correct place for these rules because they ensure correctness and consistency regardless of how the data is accessed.

3. **Emphasis on Declarative Specification:**  
   In Date’s vision, constraints should be declared in a high-level, declarative manner, not buried in triggers or procedural code. This aligns with the idea of database-wide invariants that are globally declared and enforced. While Date’s published works (such as "An Introduction to Database Systems" and the writings derived from “The Third Manifesto” with Hugh Darwen) predate many contemporary implementations, he explicitly acknowledges that integrative constraint specification should be more powerful and more pervasive than the partial set of constraints currently provided by SQL (like CHECK constraints or FOREIGN KEY constraints).

4. **The Third Manifesto Influence:**  
   In “The Third Manifesto” (by C.J. Date and Hugh Darwen), the authors advocate for a truly relational system that would support a full range of constraints, including database-wide invariants, as first-class citizens. While not discussing a specific modern-day implementation, the manifesto’s proposals for a future language (D) include robust support for constraints that apply to entire relational variables and sets of relations. This is directly in line with the concept of database-wide constraints as a logical extension of the relational model.

**Conclusion:**

While C.J. Date may not have explicitly used the term “database-wide constraints” in the modern sense (e.g., constraints that are beyond single-table CHECKs or FOREIGN KEYs, implemented via triggers or custom logic), his theoretical stance is clear: the relational model inherently calls for comprehensive integrity enforcement at the database level. Date has repeatedly criticized SQL’s failure to implement ASSERTIONS and has endorsed a model where all business rules, including those spanning multiple relations, should be enforced declaratively and centrally within the database. This strongly suggests that Date would view a well-implemented, declarative database-wide constraint mechanism as a positive development aligning with relational theory’s core principles.
</blockquote>

**Reply by danolivo** [reply]
More or less that you may find in his books. But I hope the live discussion can reveal some facts & experiences that weren't covered in papers.

**Reply by yrashk** [reply]
Agree. Need to get that meeting set up!

**Reply by ebellani** [reply]
please invite me to the discussion!

**Reply by ebellani** [reply]
I think Pascal's article is appropriate for this discussion: https://www.dbdebunk.com/2021/09/understanding-relational-constraints.html

**Reply by danolivo** [reply]
> Agree. Need to get that meeting set up!

Ok, I just wanted to say that before diving into implementing such a feature (in-core or in an extensible way), investing in a current state-of-the-art would be nice. A former Oracle developer or experienced scientist may be a good choice.

**Comment by jacobbogers** [comment]
keep Real triggers for real physical tables

An extension can (instead of creating a physical table directly) (wich it eventually needs to do, because data must live somewhere) but instead create a "virtual table" or "smart table" which can be decorated with asserts and rules that fire off whenever data that is inserted meets certain needs in a declarative way.
